File: 212-WordSearchII/212-WordSearchII.java
/*

212. Word Search II
Hard

Given a 2D board and a list of words from the dictionary, find all words in the board.

Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

Example:

Input: 
board = [
  ['o','a','a','n'],
  ['e','t','a','e'],
  ['i','h','k','r'],
  ['i','f','l','v']
]
words = ["oath","pea","eat","rain"]

Output: ["eat","oath"]

Note:

    All inputs are consist of lowercase letters a-z.
    The values of words are distinct.

*/

class Solution {
    TNode root = new TNode();
    public List<String> findWords(char[][] board, String[] words) {
        List<String> resList = new ArrayList<>();
        if(!validateInput(board, words)) return resList;
        for(String w : words)
            addToTrie(w);
        TNode temp = root;
        boolean[][] visited = new boolean[board.length][board[0].length];
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                if(temp.children.containsKey(board[i][j])){
                    dfs(board, visited, resList, temp, new StringBuilder(),i, j);
                }
            }
        }
        return resList;
    }
    private boolean validateInput(char[][] board, String[] words){
        if(words == null || board == null || words.length == 0 || board.length == 0)
            return false;
        return true;
    }
    private void dfs(char[][] board, boolean[][] visited, List<String> resList, TNode temp, StringBuilder sb, int i, int j){
        if(temp.isEnd){
            resList.add(sb.toString());
            temp.isEnd = false;
        }
        if(i < 0 || j < 0 || i >= board.length || j >= board[0].length || visited[i][j] || !temp.children.containsKey(board[i][j])) return;
        visited[i][j] = true;
        temp = temp.children.get(board[i][j]);
        sb.append(board[i][j]);
        
        dfs(board, visited, resList, temp, sb, i + 1, j);
        dfs(board, visited, resList, temp, sb, i, j + 1);
        dfs(board, visited, resList, temp, sb, i - 1, j);
        dfs(board, visited, resList, temp, sb, i, j - 1);
        
        sb.deleteCharAt(sb.length() - 1);
        visited[i][j] = false;
    }
    private void  addToTrie(String w){
        TNode temp = root;
        for(int i = 0; i < w.length(); i++){
            char ch = w.charAt(i);
            if(!temp.children.containsKey(ch))
                temp.children.put(ch, new TNode());
            temp = temp.children.get(ch);
        }
        temp.isEnd = true;
    }
}
class TNode{
    Map<Character, TNode> children;
    boolean isEnd;
    TNode(){
        children = new HashMap<>();
        isEnd = false;
    }
}

File: 212-WordSearchII/212-WordSearchII.py
"""

212. Word Search II

Given an m x n board of characters and a list of strings words, return all words on the board.

Each word must be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

Example 1:

Input: board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]], words = ["oath","pea","eat","rain"]
Output: ["eat","oath"]

Example 2:

Input: board = [["a","b"],["c","d"]], words = ["abcb"]
Output: []

Constraints:

    m == board.length
    n == board[i].length
    1 <= m, n <= 12
    board[i][j] is a lowercase English letter.
    1 <= words.length <= 3 * 104
    1 <= words[i].length <= 10
    words[i] consists of lowercase English letters.
    All the strings of words are unique.

"""

class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        def insert_in_trie(word, curr):
            for ch in word:
                if ch not in curr.children: curr.children[ch] = TrieNode()
                curr = curr.children[ch]
            curr.isEnd = True
        for word in words:
            curr = root
            insert_in_trie(word, curr)
        reslist = []
        def dfs(i, j, curr, word):
            if curr.isEnd:
                reslist.append(word)
                curr.isEnd = False
            if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]) or board[i][j] not in curr.children or vis[i][j]: return
            vis[i][j] = True
            curr = curr.children[board[i][j]]
            dfs(i + 1, j, curr, word + board[i][j])
            dfs(i, j + 1, curr, word + board[i][j])
            dfs(i - 1, j, curr, word + board[i][j])
            dfs(i, j - 1, curr, word + board[i][j])
            vis[i][j] = False

        vis = [[False] * len(board[0]) for _ in range(len(board))]
        for i in range(len(board)):
            for j in range(len(board[0])):
                curr = root
                if board[i][j] in curr.children: dfs(i, j, curr, "")
        return reslist
